### ISSUE TITLE
Improperly signed update still?

### ISSUE THREAD

**Issue by JamaicanRumCream** [sparkle-project/Sparkle#679]
I haven't updated my app in 4 years. So I did that (using Xcode 7.0, 7A220) and set the new version to be Dev signed with my ID. I downloaded Sparkle from Git and built the framework so I have v1.12.0. Exported the app, made a zip file, updated xml, upload to my web server (only http not https). Use my old version of app, update seems to go fine, it downloads the file then gives me "improperly signed update" error at the end.
I see there was some problem previously with other builds, but v1.12.0 is the newest right? What else could it be?

**Comment by pornel** [comment]
Do you have more information in the log? You should see why it's considered improperly signed.

**Comment by JamaicanRumCream** [comment]
Here's what seems relevant in that timeframe:
11/8/15 6:47:39.406 PM mdworker[3250]: code validation failed in the process of getting signing information: Error Domain=NSOSStatusErrorDomain Code=-67062 "The operation couldn’t be completed. (OSStatus error -67062.)" UserInfo=0x7fc3b404a910 {SecCSArchitecture=ppc}

11/8/15 6:47:39.692 PM mdworker[3251]: code validation failed in the process of getting signing information: Error Domain=NSOSStatusErrorDomain Code=-67062 "The operation couldn’t be completed. (OSStatus error -67062.)" UserInfo=0x7fd8d3c58ee0 {SecCSArchitecture=x86_64}

11/8/15 6:49:08.162 PM ID Font Catalog[3254]: ID Font Catalog(3254,0x11094d000) malloc: *** auto malloc[3254]: error: GC operation on unregistered thread. Thread registered implicitly. Break on auto_zone_thread_registration_error() to debug.

11/8/15 6:49:09.000 PM kernel[0]: ID Font Catalog (map: 0xffffff80470d3870) triggered DYLD shared region unnest for map: 0xffffff80470d3870, region 0x7fff95800000->0x7fff95a00000. While not abnormal for debuggers, this increases system memory footprint until the target exits.

11/8/15 6:49:48.500 PM ID Font Catalog[3254]: Sparkle Error: The update is improperly signed.

I also wanted to add that my old app did not have any signing at all IIRC, and was still using CG, if that matters.

**Comment by pornel** [comment]
That's surprisingly little. Before the line "Sparkle Error: The update is improperly signed." you should have a few more from this function:

https://github.com/sparkle-project/Sparkle/blob/0bc66ab011046ba86fba8666877cac0d102d5755/Sparkle/SUBasicUpdateDriver.m#L237

See if you get anything in ~/Library/Logs/SparkleUpdateLog.log

**Comment by pornel** [comment]
I haven't seen "error: GC operation on unregistered thread." before. AFAIK Xcode doesn't support GC at all any more. Is it possible that you still have some intermediate files/libraries compiled with an older version of Xcode?

**Comment by JamaicanRumCream** [comment]
I'm wondering if this is the error of the version running in the old app? That would be the possible CG version, which would have been the 4+ year old version of this app that is doing the actual update.

My ~/Library/Logs/SparkleUpdateLog.log has only 1 line about Transmit.app, nothing on Font Catalog.

![screen shot 2015-11-08 at 8 46 09 pm](https://cloud.githubusercontent.com/assets/5015041/11024262/e3c3c56c-8659-11e5-92ec-f7842679bf54.png)
This app is using some Applescript files, but nothing about those should be 'unusual' IMO. The only frameworks I have are Sparkle, Cocoa, ApplescriptObjC, Appkit, Coredata, Foundation.

**Comment by JamaicanRumCream** [comment]
I tried with an unsigned app and still get the same message.

**Comment by pornel** [comment]
> I'm wondering if this is the error of the version running in the old app? That would be the possible CG version, which would have been the 4+ year old version of this app that is doing the actual update.

What's the version of Sparkle that the old app is using? The old GC app would need to use 1.5b6, as there's no newer Sparkle that would run in a GC app.

I'm assuming you're only testing updates from Sparkle 1.11 or 1.12 from a new app build that uses ARC or manual memory management, is that right?

**Comment by JamaicanRumCream** [comment]
Right, the old app is using 1.5b6; I am testing/using OS X 10.10.5. My new build with ARC from a new blank project (with my good source code files copied over manually) is using 1.12.

**Comment by pornel** [comment]
In that case I'm puzzled why you're not getting a detailed reason why the signature was rejected.

"The update is improperly signed" is only in one place in the codebase:

https://github.com/sparkle-project/Sparkle/blob/0bc66ab011046ba86fba8666877cac0d102d5755/Sparkle/SUBasicUpdateDriver.m#L415

and it's preceded by a call to this function:

https://github.com/sparkle-project/Sparkle/blob/0bc66ab011046ba86fba8666877cac0d102d5755/Sparkle/SUBasicUpdateDriver.m#L237

which calls SULog (which calls NSLog and appends to the SparkleUpdateLog file) with more detailed information.

**Comment by pornel** [comment]
Maybe try running your app's update in a debugger and set a breakpoint in [validateUpdateDownloadedToPath:…](https://github.com/sparkle-project/Sparkle/blob/0bc66ab011046ba86fba8666877cac0d102d5755/Sparkle/SUBasicUpdateDriver.m#L237)?

**Comment by pornel** [comment]
If not, then just go through a checklist:

* Are both old and new apps signed with the same Developer ID?
* Do both old and new apps carry the same DSA public key?

Since 1.11 we check not only whether the update is valid, but also whether the updated will continue to have ability to verify future updates, so that you can't accidentally ship a version that will lock itself out.

**Comment by JamaicanRumCream** [comment]
It's definitely the same DSA.

**Comment by pornel** [comment]
Any luck with getting the log?

If not, I'll see if I can expose more information in the GUI.

**Comment by JamaicanRumCream** [comment]
I'm not at home with the code, it'll be till later tonight that I can try anything.

**Comment by JamaicanRumCream** [comment]
I used the updater on itself (set a build # to a lower number) and the update worked fine (running from xcode). So, I guess I will chalk this up to old versions not working well any more. Maybe there was a hiccup my signing or Dev ID along the way (combining IDs from seperate accounts into 1 recently maybe). Thanks for your time.

2015-11-09 21:11:38.785 ID Font Catalog[1164:37008] Sparkle: ===== ID Font Catalog.app =====
2015-11-09 21:11:38.786 ID Font Catalog[1164:37008] Sparkle: Extracting using '/usr/bin/ditto' '-x' '-k' '-' < '/Users/chrispav/Library/Caches/com.paveglio.IDFontCatalog/Sparkle/ID Font Catalog 4.0.0/IDFontCat4_0_0.zip' '/Users/chrispav/Library/Caches/com.paveglio.IDFontCatalog/Sparkle/ID Font Catalog 4.0.0'

**Comment by pornel** [comment]
A change of the Dev ID account could make it look like a different one. If Dev IDs don't match, then Sparkle will require signing with DSA keys instead.

If you still have the private key for the DSA public key embedded in the old app you should be able to upgrade it via DSA signing (note that Sparkle 1.5 will silently skip DSA verification if it doesn't find public key in the old app or signature in the appcast, so double-check these).
